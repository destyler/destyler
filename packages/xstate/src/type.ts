/**
 * General type utilities
 */

import type { Machine } from './machine'

export type Dict<T = any> = Record<string, T>

export type MaybeArray<T> = T | T[]

export type VoidFunction = () => void

type IfEquals<X, Y, A, B> = (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2 ? A : B

type WritableKey<T> = {
  [P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, P, never>
}[keyof T]

export type Writable<T> = Pick<T, WritableKey<T>>

type Computed<T> = Omit<T, WritableKey<T>>

/**
 * Context types
 */

/** Combined context type with readonly computed properties */
export type Context<V, C> = V & Readonly<C>

/** Computed context properties definition */
export type TComputedContext<T> = {
  [K in keyof Computed<T>]: (ctx: T) => T[K]
}

/** User-provided context (only writable properties) */
export type UserContext<TContext> = Partial<Writable<TContext>>

/** Context change listener callback */
export type ContextListener<TContext extends Dict> = (context: TContext) => void

/**
 * Event types - can be either a string or object
 */

/** Base event object interface */
export interface EventObject { type: string }

/** Event can be a type string or event object */
export type Event<TEvent extends EventObject = EventObject> = TEvent['type'] | TEvent

/** Event object with additional properties */
export interface AnyEventObject extends EventObject {
  [key: string]: any
}

/** Function to send events to the machine */
export type Send<TEvent extends EventObject = AnyEventObject> = (event: Event<TEvent>) => void

/** Event listener callback */
export type EventListener<TEvent extends EventObject = AnyEventObject> = (event: TEvent) => void

/** Extract specific event type from union */
export type ExtractEvent<TEvent extends EventObject, K> = K extends TEvent['type']
  ? Extract<TEvent, { type: K }>
  : EventObject

/**
 * Expression types - actions and activities are expressions
 */

type Expression<TContext extends Dict, TEvent extends EventObject, TReturn> = (
  context: TContext,
  event: TEvent,
) => TReturn

/** Meta information passed to expressions */
export interface Meta<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  state: State<TContext, TState>
  guards: Dict
  send: Send<TEvent>
  self: Self<TContext, TState, TEvent>
  initialContext: TContext
  initialState: string
  getState: () => State<TContext, TState, TEvent>
  getAction: (key: string) => ExpressionWithMeta<TContext, TState, TEvent, void>
  getGuard: (key: string) => GuardExpression<TContext, TState, TEvent>
}

/** Expression with access to meta information */
type ExpressionWithMeta<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject, TReturn> = (
  context: TContext,
  event: TEvent,
  meta: Meta<TContext, TState, TEvent>,
) => TReturn

/**
 * Action types
 */

/** Single action - can be string reference or function */
export type Action<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | string
  | ExpressionWithMeta<TContext, TState, TEvent, void>

/** Multiple actions - can include conditional actions */
export type Actions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | ChooseHelper<TContext, TState, TEvent>
  | MaybeArray<Action<TContext, TState, TEvent>>

/** Pure actions without conditional logic */
export type PureActions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> = MaybeArray<
  Action<TContext, TState, TEvent>
>

/** Map of action names to implementations */
export interface ActionMap<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  [action: string]: ExpressionWithMeta<TContext, TState, TEvent, void>
}

/**
 * Activity types
 */

/** Activity - long-running process with cleanup */
export type Activity<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | string
  | ExpressionWithMeta<TContext, TState, TEvent, VoidFunction | void | undefined>

/** Multiple activities */
export type Activities<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> = MaybeArray<
  Activity<TContext, TState, TEvent>
>

/** Map of activity names to implementations */
export interface ActivityMap<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  [activity: string]: ExpressionWithMeta<TContext, TState, TEvent, VoidFunction | void | undefined>
}

/**
 * Transition types
 */

/** Transition definition with target, actions, and guards */
export interface TransitionDefinition<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Target state to transition to */
  target?: TState['value'] | undefined
  /** Actions to execute during transition */
  actions?: Actions<TContext, TState, TEvent> | undefined
  /** Guard condition for the transition */
  guard?: Guard<TContext, TState, TEvent> | undefined
  /** Whether this is an internal transition */
  internal?: boolean | undefined
}

/** Delay expression function */
export type DelayExpression<TContext extends Dict, TEvent extends EventObject> = Expression<TContext, TEvent, number>

/** Delay specification - string, number, or expression */
export type Delay<TContext extends Dict, TEvent extends EventObject> =
  | string
  | number
  | DelayExpression<TContext, TEvent>

/** Map of delay names to implementations */
export interface DelayMap<TContext extends Dict, TEvent extends EventObject> {
  [delay: string]: number | DelayExpression<TContext, TEvent>
}

/** Transition with delay specification for 'after' and 'every' */
export type DelayedTransition<
  TContext extends Dict,
  TState extends StateSchema,
  TEvent extends EventObject,
> = TransitionDefinition<TContext, TState, TEvent> & {
  /** The time to delay the event, in milliseconds */
  delay?: Delay<TContext, TEvent> | undefined
}

/** Collection of delayed transitions */
export type DelayedTransitions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | Record<string | number, TState['value'] | MaybeArray<TransitionDefinition<TContext, TState, TEvent>>>
  | Array<DelayedTransition<TContext, TState, TEvent>>

/** Transition can be a string (target) or full definition object */
export type Transition<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | TState['value']
  | TransitionDefinition<TContext, TState, TEvent>

/** Multiple transition possibilities with guards */
export type Transitions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | Transition<TContext, TState, TEvent>
  | Array<TransitionDefinition<TContext, TState, TEvent>>

/** Map of event types to their transitions */
export type TransitionDefinitionMap<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> = {
  [K in TEvent['type']]?:
    | TState['value']
    | MaybeArray<TransitionDefinition<TContext, TState, ExtractEvent<TEvent, K>>>
    | undefined
}

/**
 * State node definition
 */

/** Configuration for a single state node */
export interface StateNode<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** The type of this state node */
  type?: 'final' | undefined
  /** The tags for the state node */
  tags?: MaybeArray<TState['tags'] extends string ? TState['tags'] : string> | undefined
  /** Activities to start on entry and stop on exit */
  activities?: Activities<TContext, TState, TEvent> | undefined
  /** Event to transition mappings */
  on?: TransitionDefinitionMap<TContext, TState, TEvent> | undefined
  /** Actions to execute on state entry */
  entry?: Actions<TContext, TState, TEvent> | undefined
  /** Actions to execute on state exit */
  exit?: Actions<TContext, TState, TEvent> | undefined
  /** Meta data associated with this state node */
  meta?: string | Dict | undefined
  /** Delayed transitions using setTimeout */
  after?: DelayedTransitions<TContext, TState, TEvent> | undefined
  /** Interval-based actions using setInterval */
  every?:
    | Record<string | number, Actions<TContext, TState, TEvent>>
    | Array<{
      delay?: number | string | Expression<TContext, TEvent, number>
      actions: Actions<TContext, TState, TEvent>
      guard?: Guard<TContext, TState, TEvent>
    }>
    | undefined
}

/**
 * Guard types
 */

/** Meta information available to guard functions */
export interface GuardMeta<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  state: Pick<State<TContext, TState, TEvent>, 'matches'>
}

/** Guard expression function */
export type GuardExpression<
  TContext extends Dict,
  TState extends StateSchema,
  TEvent extends EventObject,
  TReturn = boolean,
> = (context: TContext, event: TEvent, guardMeta: GuardMeta<TContext, TState, TEvent>) => TReturn

/** Guard helper for complex guard logic */
export interface GuardHelper<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  predicate: (guards: Dict) => GuardExpression<TContext, TState, TEvent>
}

/** Helper for conditional action selection */
export interface ChooseHelper<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  predicate: (
    guards: Dict,
  ) => GuardExpression<TContext, TState, TEvent, PureActions<TContext, TState, TEvent> | undefined>
}

/** Guard can be string reference, function, or helper */
export type Guard<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> =
  | string
  | GuardExpression<TContext, TState, TEvent>
  | GuardHelper<TContext, TState, TEvent>

/** Map of guard names to implementations */
export interface GuardMap<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  [guard: string]: GuardExpression<TContext, TState, TEvent>
}

/**
 * State types
 */

/** Schema defining state structure */
export interface StateSchema {
  value: string
  tags?: string | undefined
}

/** State initialization object */
export interface StateInitObject<TContext, TState extends StateSchema> {
  context?: TContext | undefined
  value: TState['value'] | null
  tags?: TState['tags'][] | undefined
}

/** State initialization type */
export type StateInit<TContext, TState extends StateSchema> = StateInitObject<TContext, TState>

/** State change listener callback */
export type StateListener<
  TContext extends Dict,
  TState extends StateSchema,
  TEvent extends EventObject = EventObject,
> = (state: State<TContext, TState, TEvent>) => void

/** Information about state transition */
export interface StateInfo<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Whether the state was re-entered */
  reenter: boolean
  /** Whether the state changed */
  changed: boolean
  /** The transition definition that caused this state */
  transition: TransitionDefinition<TContext, TState, TEvent> | undefined
  /** The state node configuration */
  stateNode: StateNode<TContext, TState, TEvent> | undefined
  /** The target state value */
  target: TState['value']
}

/**
 * Machine configuration types
 */

/** Main machine configuration */
export interface MachineConfig<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Actions to run after machine instantiation */
  created?: Actions<TContext, TState, TEvent> | undefined
  /** Actions to run when machine starts */
  entry?: Actions<TContext, TState, TEvent> | undefined
  /** Actions to run when machine stops */
  exit?: Actions<TContext, TState, TEvent> | undefined
  /** Root level activities */
  activities?: Activities<TContext, TState, TEvent> | undefined
  /** Unique machine identifier */
  id?: string | undefined
  /** Extended state data */
  context?: Writable<TContext> | undefined
  /** Context change watchers */
  watch?: { [K in keyof TContext]?: Actions<TContext, TState, TEvent> } | undefined
  /** Computed properties */
  computed?: Partial<TComputedContext<TContext>> | undefined
  /** Initial state */
  initial?: TState['value'] | undefined
  /** State node configurations */
  states?: Partial<Record<TState['value'], StateNode<TContext, TState, TEvent>>> | undefined
  /** Root level event transitions */
  on?: TransitionDefinitionMap<TContext, TState, TEvent> | undefined
}

/** Current state representation */
export interface State<
  TContext extends Dict,
  TState extends StateSchema = StateSchema,
  TEvent extends EventObject = AnyEventObject,
> {
  /** Current state value */
  value: TState['value'] | null
  /** Previous state value */
  previousValue: TState['value'] | null
  /** Current event */
  event: TEvent
  /** Previous event */
  previousEvent: TEvent
  /** Current context */
  context: TContext
  /** Whether machine is in final state */
  done: boolean
  /** Check if event can be sent */
  can: (event: string) => boolean
  /** Check if state matches given values */
  matches: (...value: TState['value'][]) => boolean
  /** Check if state has given tag */
  hasTag: (value: TState['tags']) => boolean
  /** Events that can be sent from current state */
  nextEvents: string[]
  /** Whether state has changed */
  changed: boolean
  /** Current state tags */
  tags: TState['tags'][]
}

/**
 * Machine options types
 */

/** Options for machine configuration */
export interface MachineOptions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Enable debug logging */
  debug?: boolean | undefined
  /** Guard implementations */
  guards?: GuardMap<TContext, TState, TEvent> | undefined
  /** Action implementations */
  actions?: ActionMap<TContext, TState, TEvent> | undefined
  /** Delay implementations */
  delays?: DelayMap<TContext, TEvent> | undefined
  /** Activity implementations */
  activities?: ActivityMap<TContext, TState, TEvent> | undefined
  /** Synchronous execution mode */
  sync?: boolean | undefined
  /** Custom comparison functions for context properties */
  compareFns?: { [K in keyof TContext]?: CompareFn<TContext[K]> } | undefined
}

/** Options for machine hooks */
export interface HookOptions<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Synchronous execution mode */
  sync?: boolean
  /** Action implementations */
  actions?: ActionMap<TContext, TState, TEvent> | undefined
  /** Initial state */
  state?: StateInit<TContext, TState> | undefined
  /** Initial context */
  context?: UserContext<TContext> | undefined
}

/** Machine self-reference for advanced operations */
export interface Self<TContext extends Dict, TState extends StateSchema, TEvent extends EventObject> {
  /** Machine identifier */
  id: string
  /** Send event to this machine */
  send: (event: Event<TEvent>) => void
  /** Send event to parent machine */
  sendParent: (evt: AnyEventObject) => void
  /** Send event to child machine */
  sendChild: (evt: Event<TEvent>, to: string | ((ctx: TContext) => string)) => void
  /** Stop this machine */
  stop: VoidFunction
  /** Stop child machine */
  stopChild: (id: string) => void
  /** Stop activity */
  stopActivity: (id: string) => void
  /** Spawn new actor */
  spawn: <T>(src: T | (() => T), id?: string) => T
  /** Current state */
  state: State<TContext, TState, TEvent>
  /** Initial context */
  initialContext: TContext
  /** Initial state */
  initialState: string
}

/** Machine status enumeration */
export enum MachineStatus {
  NotStarted = 'Not Started',
  Running = 'Running',
  Stopped = 'Stopped',
}

/** Built-in action types */
export enum ActionTypes {
  Start = 'machine.start',
  Stop = 'machine.stop',
  Created = 'machine.created',
  Init = 'machine.init',
}

/** Machine type enumeration */
export enum MachineType {
  Machine = 'machine',
  Actor = 'machine.actor',
}

/** Comparison function type */
export type CompareFn<T = any> = (prev: T, next: T) => boolean

// utils

type AnyFunction = (...args: any[]) => any
type ReturnTypeOrValue<T> = T extends AnyFunction ? ReturnType<T> : T

export type StateFrom<T> =
  ReturnTypeOrValue<T> extends infer R
    ? R extends Machine<infer TContext, infer TState, infer TEvent>
      ? State<TContext, TState, TEvent>
      : never
    : never

export type ContextFrom<T> =
  ReturnTypeOrValue<T> extends infer R ? (R extends Machine<infer TContext, any, any> ? TContext : never) : never

export type EventFrom<T> =
  ReturnTypeOrValue<T> extends infer R ? (R extends Machine<any, any, infer TEvent> ? TEvent : never) : never

export interface ContextRef<T> { current: T | undefined }
