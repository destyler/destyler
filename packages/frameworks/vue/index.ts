export * from './src/composition/actor'
export * from './src/composition/machine'
export * from './src/composition/snapshot'
export * from './src/shared/normalize-props'
export { mergeProps } from '@zag-js/core'
export type { ContextFrom, EventFrom, StateFrom } from '@zag-js/core'
