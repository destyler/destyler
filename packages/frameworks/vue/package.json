{
  "name": "@destyler/vue",
  "type": "module",
  "version": "0.0.7-beta.7",
  "description": "the vue wrapper for destyler",
  "author": "Elone Hoo <elonehoo@gmail.com>",
  "license": "MIT",
  "exports": {
    ".": "./dist/index.js"
  },
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "*.d.ts",
    "dist"
  ],
  "scripts": {
    "clear": "rimraf ./dist",
    "clear:node_modules": "rimraf ./node_modules",
    "prepublishOnly": "pnpm run build",
    "build": "pnpm run clear && tsdown",
    "stub": "tsdown --watch"
  },
  "peerDependencies": {
    "vue": ">=3.0.0"
  },
  "dependencies": {
    "@destyler/store": "workspace:*",
    "@destyler/types": "workspace:*",
    "@destyler/xstate": "workspace:*"
  },
  "devDependencies": {
    "@destyler/utils": "workspace:*",
    "vue": "catalog:frontend"
  }
}
