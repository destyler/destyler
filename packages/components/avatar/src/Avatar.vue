<script setup lang="ts">
import { onMounted, ref } from 'vue'
import type { DestylerAvatarProps } from './props'

defineOptions({
  name: 'DestylerAvatar',
})

const { src, alt, delayMs } = withDefaults(defineProps<DestylerAvatarProps>(), {
  delayMs: 600,
})

const canRender = ref(false)
let timeout: ReturnType<typeof setTimeout> | undefined

if (delayMs) {
  timeout = setTimeout(() => {
    canRender.value = true
    clearTimeout(timeout)
  }, delayMs)
}
else {
  canRender.value = true
}

onMounted(() => {
  imageInit(src!)
})

const imageLoadingStatus = ref<'loaded' | 'error'>()
function imageInit(src: string) {
  const image = new Image()
  image.src = src
  image.onload = () => {
    imageLoadingStatus.value = 'loaded'
  }
  image.onerror = () => {
    imageLoadingStatus.value = 'error'
  }
}
</script>

<template>
  <span
    destyler-avatar-root
  >
    <img
      v-if="imageLoadingStatus === 'loaded'"
      destyler-avatar-img
      :src="src"
      :alt="alt"
    >
    <span
      v-if="imageLoadingStatus !== 'loaded' && canRender"
      destyler-avatar-fallback
    >
      {{ fallback }}
    </span>
  </span>
</template>
