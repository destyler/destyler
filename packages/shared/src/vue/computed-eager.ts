import type { Ref, WatchOptionsBase } from 'vue'
import { readonly, shallowRef, watchEffect } from 'vue'

/**
 * @param fn effect function
 * @param options WatchOptionsBase
 * @returns readonly ref
 */
export function computedEager<T>(fn: () => T, options?: WatchOptionsBase): Readonly<Ref<T>> {
  const result = shallowRef()

  watchEffect(() => {
    result.value = fn()
  }, {
    ...options,
    flush: options?.flush ?? 'sync',
  })

  return readonly(result)
}

// alias
export { computedEager as eagerComputed }
