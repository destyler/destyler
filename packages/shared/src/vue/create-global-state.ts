import { effectScope } from 'vue'
import type { AnyFn } from './types'

/**
 * Keep states in the global scope to be reusable across Vue instances.
 *
 * @param stateFactory A factory function to create the state
 */
export function createGlobalState<Fn extends AnyFn>(
  stateFactory: Fn,
): Fn {
  let initialized = false
  let state: any
  const scope = effectScope(true)

  return ((...args: any[]) => {
    if (!initialized) {
      state = scope.run(() => stateFactory(...args))!
      initialized = true
    }
    return state
  }) as Fn
}
