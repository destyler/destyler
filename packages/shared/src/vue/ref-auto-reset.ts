import type { Ref } from 'vue'
import { customRef } from 'vue'
import type { MaybeRefOrGetter } from './types'
import { toValue } from './to-value'
import { tryOnScopeDispose } from './try-on-scope-dispose'

export function refAutoReset<T>(defaultValue: MaybeRefOrGetter<T>, afterMs: MaybeRefOrGetter<number> = 10000): Ref<T> {
  return customRef<T>((track, trigger) => {
    let value: T = toValue(defaultValue)
    let timer: any

    const resetAfter = () =>
      setTimeout(() => {
        value = toValue(defaultValue)
        trigger()
      }, toValue(afterMs))

    tryOnScopeDispose(() => {
      clearTimeout(timer)
    })

    return {
      get() {
        track()
        return value
      },
      set(newValue) {
        value = newValue
        trigger()

        clearTimeout(timer)
        timer = resetAfter()
      },
    }
  })
}
