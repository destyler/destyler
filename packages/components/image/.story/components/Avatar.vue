<script setup lang="ts">
import { Image, ImageFallback, ImageRoot } from '../../src'

defineProps<{ delay?: number }>()
</script>

<template>
  <ImageRoot
    class="bg-blackA3 inline-flex h-[45px] w-[45px] select-none items-center justify-center overflow-hidden rounded-full align-middle"
  >
    <Image
      class="h-full w-full rounded-[inherit] object-cover"
      src="https://github.com/elonehoo.png"
      alt="Elone Hoo"
    />
    <ImageFallback
      class="text-green11 leading-1 flex h-full w-full items-center justify-center bg-white text-[15px] font-medium"
      :delay-ms="delay"
    >
      EH
    </ImageFallback>
  </ImageRoot>
</template>
