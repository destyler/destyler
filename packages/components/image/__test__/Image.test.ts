import { beforeAll, describe, expect, it } from 'vitest'
import { axe } from 'vitest-axe'
import type { VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { nextTick } from 'vue'
import { findByAltText, findByText } from '@testing-library/vue'
import Avatar from './Image.spec.vue'

const FALLBACK = 'EH'
const DELAY = 350

it('should pass axe accessibility tests', async () => {
  const wrapper = mount(Avatar)
  expect(await axe(wrapper.element)).toHaveNoViolations()
})

describe('given an Avatar with fallback and a working image', async () => {
  let wrapper: VueWrapper<InstanceType<typeof Avatar>>
  beforeAll(async () => {
    (window.Image as any) = class MockImage {
      onload: () => void = () => {}
      src = ''
      constructor() {
        setTimeout(() => {
          this.onload()
        }, DELAY)
        return this
      }
    }
    wrapper = mount(Avatar)
  })

  it('should render the fallback initially', async () => {
    await nextTick()
    expect(wrapper.html()).includes(FALLBACK)
  })

  it('should have alt text on the image', async () => {
    const image = await findByAltText(wrapper.element as HTMLElement, 'Elone Hoo')
    expect(image).toBeTruthy()
  })

  it('should match before image loaded snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('given an Avatar with fallback and delayed render', () => {
  let wrapper: VueWrapper<InstanceType<typeof Avatar>>

  beforeAll(async () => {
    (window.Image as any) = class MockImage {
      onload: () => void = () => {}
      src = ''
      constructor() {
        setTimeout(() => {
          this.onload()
        }, DELAY)
        return this
      }
    }
    wrapper = mount(Avatar, {
      props: {
        delay: 300,
      },
    })
  })

  it('should not render a fallback immediately', async () => {
    expect(wrapper.text()).not.toContain(FALLBACK)
  })

  it('should render a fallback after the delay', async () => {
    expect(wrapper.text()).not.toContain(FALLBACK)
    const fallback = await findByText(wrapper.element as HTMLElement, FALLBACK)
    expect(wrapper.element).toContain(fallback)
  })
})
