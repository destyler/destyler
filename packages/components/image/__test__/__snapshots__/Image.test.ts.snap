// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given an Avatar with fallback and a working image > should match before image loaded snapshot 1`] = `"<span><img role="img" src="https://github.com/elonehoo.png" alt="Elone Hoo"><!----></span>"`;
